analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/SymbolicCheckListTest.cs

```diff
@@ -0,0 +1,111 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SonarAnalyzer.SymbolicExecution.Roslyn;
+using SonarAnalyzer.UnitTest.TestFramework.SymbolicExecution;
+
+namespace SonarAnalyzer.UnitTest.SymbolicExecution.Roslyn
+{
+    [TestClass]
+    public class SymbolicCheckListTest
+    {
+        [TestMethod]
+        public void Constructor_Null_Throws() =>
+            ((Func<SymbolicCheckList>)(() => new SymbolicCheckList(null))).Should().Throw<ArgumentNullException>().And.ParamName.Should().Be("checks");
+
+        [TestMethod]
+        public void Notifications_ExecutedForAll()
+        {
+            var a = new FlagTestCheck();
+            var b = new FlagTestCheck();
+            var context = new SymbolicContext(new(), null, ProgramState.Empty);
+            var sut = new SymbolicCheckList(new[] { a, b });
+
+            a.WasExitReached.Should().BeFalse();
+            b.WasExitReached.Should().BeFalse();
+            sut.ExitReached(context);
+            a.WasExitReached.Should().BeTrue();
+            b.WasExitReached.Should().BeTrue();
+
+            a.WasExecutionCompleted.Should().BeFalse();
+            b.WasExecutionCompleted.Should().BeFalse();
+            sut.ExecutionCompleted();
+            a.WasExecutionCompleted.Should().BeTrue();
+            b.WasExecutionCompleted.Should().BeTrue();
+
+            a.WasPreProcessed.Should().BeFalse();
+            b.WasPreProcessed.Should().BeFalse();
+            sut.PreProcess(context);
+            a.WasPreProcessed.Should().BeTrue();
+            b.WasPreProcessed.Should().BeTrue();
+
+            a.WasPostProcessed.Should().BeFalse();
+            b.WasPostProcessed.Should().BeFalse();
+            sut.PostProcess(context);
+            a.WasPostProcessed.Should().BeTrue();
+            b.WasPostProcessed.Should().BeTrue();
+        }
+
+        [TestMethod]
+        public void PostProcess_CanReturnMultipleStates()
+        {
+            var triple = new PostProcessTestCheck(x => new[] { x.State, x.State, x.State });
+            var sut = new SymbolicCheckList(new[] { triple, triple });
+            sut.PostProcess(new(new(), null, ProgramState.Empty)).Should().HaveCount(9);
+        }
+
+        [TestMethod]
+        public void PostProcess_CanReturnNoStates()
+        {
+            var empty = new PostProcessTestCheck(x => Array.Empty<ProgramState>());
+            var sut = new SymbolicCheckList(new[] { empty });
+            sut.PostProcess(new(new(), null, ProgramState.Empty)).Should().HaveCount(0);
+        }
+
+        private class FlagTestCheck : SymbolicCheck
```

Comment:
Moq or NSubstitute could be used  to check for method invocations, and in that case this class could be removed.